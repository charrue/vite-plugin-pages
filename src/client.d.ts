declare module 'vite-plugin-pages/client' {
  import { RouteRecordRaw } from 'vue-router'
  declare const routes: RouteRecordRaw[]
  export default routes
}
