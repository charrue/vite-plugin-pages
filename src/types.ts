/**
 * https://github.com/brattonross/vite-plugin-voie/blob/main/packages/vite-plugin-voie/src/options.ts
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file at
 * https://github.com/brattonross/vite-plugin-voie/blob/main/LICENSE
 */

export type ImportMode = 'sync' | 'async'
export type ImportModeResolveFn = (filepath: string) => ImportMode

export interface Route {
  name?: string
  path: string
  props?: boolean
  component: string
  children?: Route[]
  meta?: Record<string, unknown>
}

export interface PageDirOptions {
  dir: string
  baseRoute: string
}

/**
 * Plugin options.
 */
interface Options {
  /**
   * Relative path to the directory to search for page components.
   * @default 'src/pages'
   */
  pagesDir: string | string[] | PageDirOptions[]
  /**
   * Valid file extensions for page components.
   * @default ['vue', 'js']
   */
  extensions: string[]
  /**
   * List of path globs to exclude when resolving pages.
   */
  exclude: string[]
  /**
   * Import routes directly or as async components
   * @default 'async'
   */
  importMode: ImportMode | ImportModeResolveFn
  /**
   * Sync load top level index file
   * @default true
   */
  syncIndex: boolean
  /**
   * Set default route block parser, or use `<route lang=xxx>` in SFC route block
   * @default 'json5'
   */
  routeBlockLang: 'json5' | 'json' | 'yaml'
  /**
   * Extend route records
   */
  extendRoute?: (route: Route, parent: Route | undefined) => Route | void
}

export type UserOptions = Partial<Options>

export interface ResolvedOptions extends Options {
  /**
   * Resolves to the `root` value from Vite config.
   * @default config.root
   */
  root: string

  /**
   * Page Dir as a normalized array of PageDirOptions
   */
  pagesDirOptions: PageDirOptions[]
}
