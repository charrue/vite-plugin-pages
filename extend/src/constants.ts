export const MACROS_KEY = 'definePageRoute'
export const MACROS_KEY_RE = /definePageRoute\(([\s\S]*?)\);?/g
