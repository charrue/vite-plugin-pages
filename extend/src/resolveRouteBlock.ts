export const MACROS_KEY = 'definePageRoute'
export const MACROS_KEY_RE = /definePageRoute\(([\s\S]*?)\);?/g
export const MACROS_FN_STR = `function ${MACROS_KEY}(config) { return config };`

export const isJsx = (filePath: string) => {
  return filePath.endsWith('.jsx') || filePath.endsWith('.tsx')
}

export const resolveRouteBlock = (content: string, filePath: string) => {
  if (isJsx(filePath) && content.includes(MACROS_KEY)) {
    const body = content.match(MACROS_KEY_RE)?.[0]
    if (body) {
      // eslint-disable-next-line no-new-func
      const routeConfig = new Function(`${MACROS_FN_STR} return ${body}`)()
      return routeConfig
    }
  }
}
