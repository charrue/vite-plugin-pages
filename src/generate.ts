/**
 * https://github.com/brattonross/vite-plugin-voie/blob/main/packages/vite-plugin-voie/src/routes.ts
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file at
 * https://github.com/brattonross/vite-plugin-voie/blob/main/LICENSE
 */

import { resolve, parse, join } from 'path'
import fg from 'fast-glob'
import chalk from 'chalk'
import { Route, Options } from './types'
import {
  normalizePath,
  debug,
  isDynamicRoute,
  resolveImportMode,
  pathToName,
} from './utils'
import { stringifyRoutes } from './stringify'

function prepareRoutes(
  routes: Route[],
  options: Options,
  parent?: Route,
) {
  for (const route of routes) {
    if (route.name)
      route.name = route.name.replace(/-index$/, '')

    if (parent)
      route.path = route.path.replace(/^\//, '').replace(/\?$/, '')

    route.props = true

    if (route.children) {
      delete route.name
      route.children = prepareRoutes(route.children, options, route)
    }

    if (typeof options.extendRoute === 'function')
      Object.assign(route, options.extendRoute(route, parent) || {})
  }
  return routes
}

export function generateRoutes(filesPath: string[], options: Options): string {
  const {
    root,
    pagesDir,
    extensions,
    exclude,
  } = options
  const cwd = normalizePath(resolve(root, pagesDir))
  const extensionsRE = new RegExp(`\\.(${extensions.join('|')})$`)

  const routes: Route[] = []

  for (const filePath of filesPath) {
    const resolvedPath = filePath.replace(extensionsRE, '')
    const pathNodes = resolvedPath.split('/')

    const component = `/${pagesDir}/${filePath}`
    const route: Route = {
      name: '',
      path: '',
      component,
    }

    let parentRoutes = routes

    for (let i = 0; i < pathNodes.length; i++) {
      const node = pathNodes[i]
      const isDynamic = isDynamicRoute(node)
      const isLastOne = i === pathNodes.length - 1
      const normalizedPart = (isDynamic
        ? node.replace(/^\[(\.{3})?/, '').replace(/\]$/, '')
        : node
      ).toLowerCase()

      route.name += route.name ? `-${normalizedPart}` : normalizedPart

      // Check nested route
      const parent = parentRoutes.find(node => node.name === route.name)

      if (parent) {
        parent.children = parent.children || []
        parentRoutes = parent.children
        route.path = ''
      }
      else if (normalizedPart === 'index' && !route.path) {
        route.path += '/'
      }
      else if (normalizedPart !== 'index') {
        if (isDynamic) {
          route.path += `/:${normalizedPart}`
          // Catch-all route
          if (/^\[\.{3}/.test(node))
            route.path += '(.*)'
          else if (isLastOne)
            route.path += '?'
        }
        else {
          route.path += `/${normalizedPart}`
        }
      }
    }
    parentRoutes.push(route)
  }

  const preparedRoutes = prepareRoutes(routes, options)

  return preparedRoutes
}

export function generateClientCode(filesPath: string[], options: Options) {
  const routes = generateRoutes(filesPath, options)
  const { imports, stringRoutes } = stringifyRoutes(routes, options)

  return `${imports.join('\n')}\n\nconst routes = ${stringRoutes}\n\nexport default routes`
}
