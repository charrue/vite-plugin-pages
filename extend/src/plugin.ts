import { MACROS_FN_STR, MACROS_KEY, isJsx } from './resolveRouteBlock'
import type { Plugin } from 'vite'

export function definePageRoutePlugin(): Plugin {
  return {
    name: 'charrue:definePageRoute',
    enforce: 'pre',

    async transform(code, id) {
      const [filepath] = id.split('?')

      if (isJsx(filepath)) {
        if (!code.includes(MACROS_KEY)) return code
        return `${MACROS_FN_STR}\n${code}`
      }

      return code
    },
  }
}
