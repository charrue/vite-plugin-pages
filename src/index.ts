import { resolve } from 'path'
import type { Plugin, ResolvedConfig, ModuleNode } from 'vite'
import { Route, ResolvedOptions, UserOptions, PageDirOptions } from './types'
import { getFilesFromPath } from './files'
import { generateRoutes, generateClientCode, updateRouteFromHMR } from './generate'
import { debug, normalizePath } from './utils'
import { parseVueRequest } from './query'

const ID = 'pages-generated'

function resolveOptions(userOptions: UserOptions): ResolvedOptions {
  const {
    pagesDir = ['src/pages'],
    extensions = ['vue', 'js'],
    importMode = 'async',
    routeBlockLang = 'json5',
    exclude = [],
    syncIndex = true,
  } = userOptions

  const root = process.cwd()
  const pagesDirPaths: string[] = []

  let pagesDirOptions: PageDirOptions[] = []

  if (typeof pagesDir === 'string') {
    pagesDirOptions = pagesDirOptions.concat({ dir: pagesDir, baseRoute: '' })
  }
  else {
    for (const dir of pagesDir) {
      if (typeof dir === 'string')
        pagesDirOptions = pagesDirOptions.concat({ dir, baseRoute: '' })
      else if (dir as PageDirOptions)
        pagesDirOptions = pagesDirOptions.concat(dir)
    }
  }

  return Object.assign(
    {},
    {
      routeBlockLang,
      root,
      pagesDir,
      pagesDirPaths,
      pagesDirOptions,
      extensions,
      importMode,
      exclude,
      syncIndex,
    },
    userOptions,
  )
}

function routePlugin(userOptions: UserOptions = {}): Plugin {
  let config: ResolvedConfig | undefined
  let filesPath: string[] = []
  let generatedRoutes: Route[] | null | undefined
  let pagesDirPaths: string[] = []

  const options: ResolvedOptions = resolveOptions(userOptions)

  return {
    name: 'vite-plugin-pages',
    enforce: 'pre',
    configResolved(_config) {
      config = _config
      options.root = config.root
    },
    resolveId(id) {
      if (id === ID)
        return ID
    },
    async load(id) {
      if (id === ID) {
        debug('Loading...')
        debug('Loading Gen Routes: %O', generatedRoutes)

        if (!generatedRoutes) {
          generatedRoutes = []
          filesPath = []
          pagesDirPaths = []
          for (const pageDir of options.pagesDirOptions) {
            const pageDirPath = normalizePath(resolve(options.root, pageDir.dir))
            pagesDirPaths = pagesDirPaths.concat(pageDirPath)
            debug('Loading PageDir: %O', pageDirPath)

            const files = await getFilesFromPath(pageDirPath, options)
            filesPath = filesPath.concat(files.map(f => `${pageDirPath}/${f}`))
            debug('FilesPath: %O', files)

            const routes = generateRoutes(files, pageDir, pageDirPath, options)
            generatedRoutes = generatedRoutes.concat(routes)
            // debug('Routes: %O', generatedRoutes)
          }
        }

        const clientCode = generateClientCode(generatedRoutes, options)

        // debug('Client code: %O', clientCode)

        return clientCode
      }
    },
    async transform(code: string, id: string) {
      const { query } = parseVueRequest(id)

      if (query && query.vue && query.type === 'route') {
        return {
          code: 'export default {}',
          map: null,
        }
      }

      return {
        code,
        map: null,
      }
    },
    async handleHotUpdate({ file, server, read }) {
      const extensionsRE = new RegExp(`\\.(${options.extensions.join('|')})$`)
      const matchedPath = (pagesDirPaths).find(p => file.startsWith(`${p}/`))
      // Handle pages HMR
      if (matchedPath && extensionsRE.test(file)) {
        let needReload = false

        // HMR on new file created
        // Otherwise, handle HMR from custom block
        if (!filesPath.includes(file)) {
          generatedRoutes = null
          needReload = true
        }
        else if (generatedRoutes) {
          const content = await read()
          needReload = updateRouteFromHMR(content, file, generatedRoutes, options)
        }

        if (needReload) {
          const { moduleGraph } = server
          const module = moduleGraph.getModuleById(ID)

          debug('Reload for file: %s', file.replace(options.root, ''))

          return [module] as ModuleNode[]
        }
      }
    },
  }
}

export * from './types'
export default routePlugin
