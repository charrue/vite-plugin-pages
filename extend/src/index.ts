import { definePageRoutePlugin } from './plugin'
import { resolveRouteBlock } from './resolveRouteBlock'

export { definePageRoutePlugin, resolveRouteBlock }
