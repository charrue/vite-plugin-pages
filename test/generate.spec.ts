import { generateRoutes, generateClientCode } from '../src/generate'
import { resolveOptions } from '../src/index'
import { Route } from '../src/types'

const options = resolveOptions({
  pagesDir: 'test/assets/pages',
})
const nuxtOptions = resolveOptions({
  pagesDir: 'test/assets/nuxt-pages',
  nuxtStyle: true,
})
const files = ['[...all].vue', 'home.vue', 'index.vue', 'user.vue', 'user/[id].vue']
const nuxtFiles = ['_.vue', 'home.vue', 'index.vue', 'user.vue', 'user/_id.vue']
const pageDir = options.pagesDirOptions[0]
const nuxtPageDir = nuxtOptions.pagesDirOptions[0]

const expectRoutes = [
  {
    component: '/test/assets/pages/[...all].vue',
    name: 'all',
    path: '/:all(.*)',
    props: true,
  },
  {
    name: 'home',
    path: '/home',
    component: '/test/assets/pages/home.vue',
    props: true,
  },
  {
    name: 'index',
    path: '/',
    component: '/test/assets/pages/index.vue',
    props: true,
  },
  {
    path: '/user',
    component: '/test/assets/pages/user.vue',
    children: [
      {
        name: 'name-override',
        path: ':id',
        component: '/test/assets/pages/user/[id].vue',
        props: true,
        meta: {
          requiresAuth: false,
        },
      },
    ],
    props: true,
  },
]

describe('Generate', () => {
  let routes: Route[] = []
  test('Routes', () => {
    routes = generateRoutes(files, pageDir, options)
    expect(routes).toMatchSnapshot('routes')
  })
  test('Nuxt Style Routes', () => {
    routes = generateRoutes(nuxtFiles, nuxtPageDir, nuxtOptions)
    expect(routes).toMatchSnapshot('nuxt style routes')
  })
  test('Client Code', () => {
    const code = generateClientCode(expectRoutes, options)
    expect(code).toMatchSnapshot('client code')
  })
})
