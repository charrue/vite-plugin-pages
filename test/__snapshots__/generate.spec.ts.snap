// Vitest Snapshot v1

exports[`Generate > Nuxt Style Routes > client code 1`] = `
"import _examples_nuxt_style_src_pages_index_vue from \\"/examples/nuxt-style/src/pages/index.vue\\";

const routes = [{\\"path\\":\\"/about\\",\\"component\\":() => import('/examples/nuxt-style/src/pages/about.vue'),\\"children\\":[{\\"name\\":\\"about\\",\\"path\\":\\"\\",\\"component\\":() => import('/examples/nuxt-style/src/pages/about/index.vue'),\\"props\\":true},{\\"name\\":\\"about-id\\",\\"path\\":\\":id\\",\\"component\\":() => import('/examples/nuxt-style/src/pages/about/_id.vue'),\\"props\\":true},{\\"name\\":\\"who-me-override\\",\\"path\\":\\"who/me\\",\\"component\\":() => import('/examples/nuxt-style/src/pages/about/who/me.vue'),\\"props\\":true,\\"meta\\":{\\"requiresAuth\\":false}}],\\"props\\":true},{\\"name\\":\\"components\\",\\"path\\":\\"/components\\",\\"component\\":() => import('/examples/nuxt-style/src/pages/components.vue'),\\"props\\":true},{\\"name\\":\\"name-override\\",\\"path\\":\\"/\\",\\"component\\":_examples_nuxt_style_src_pages_index_vue,\\"props\\":true,\\"meta\\":{\\"requiresAuth\\":false}},{\\"name\\":\\"all\\",\\"path\\":\\"/:all(.*)*\\",\\"component\\":() => import('/examples/nuxt-style/src/pages/_.vue'),\\"props\\":true},{\\"name\\":\\"id\\",\\"path\\":\\"/:id\\",\\"component\\":() => import('/examples/nuxt-style/src/pages/_id.vue'),\\"props\\":true},{\\"name\\":\\"sensor-current\\",\\"path\\":\\"/:sensor/current\\",\\"component\\":() => import('/examples/nuxt-style/src/pages/_sensor/current.vue'),\\"props\\":true}];

export default routes;"
`;

exports[`Generate > Nuxt Style Routes > routes 1`] = `
[
  {
    "children": [
      {
        "component": "/examples/nuxt-style/src/pages/about/index.vue",
        "customBlock": undefined,
        "name": "about",
        "path": "",
        "props": true,
      },
      {
        "component": "/examples/nuxt-style/src/pages/about/_id.vue",
        "customBlock": undefined,
        "name": "about-id",
        "path": ":id",
        "props": true,
      },
      {
        "component": "/examples/nuxt-style/src/pages/about/who/me.vue",
        "meta": {
          "requiresAuth": false,
        },
        "name": "who-me-override",
        "path": "who/me",
        "props": true,
      },
    ],
    "component": "/examples/nuxt-style/src/pages/about.vue",
    "customBlock": undefined,
    "path": "/about",
    "props": true,
  },
  {
    "component": "/examples/nuxt-style/src/pages/components.vue",
    "customBlock": undefined,
    "name": "components",
    "path": "/components",
    "props": true,
  },
  {
    "component": "/examples/nuxt-style/src/pages/index.vue",
    "meta": {
      "requiresAuth": false,
    },
    "name": "name-override",
    "path": "/",
    "props": true,
  },
  {
    "component": "/examples/nuxt-style/src/pages/_.vue",
    "customBlock": undefined,
    "name": "all",
    "path": "/:all(.*)*",
    "props": true,
  },
  {
    "component": "/examples/nuxt-style/src/pages/_id.vue",
    "customBlock": undefined,
    "name": "id",
    "path": "/:id",
    "props": true,
  },
  {
    "component": "/examples/nuxt-style/src/pages/_sensor/current.vue",
    "customBlock": undefined,
    "name": "sensor-current",
    "path": "/:sensor/current",
    "props": true,
  },
]
`;

exports[`Generate > React routes > client code 1`] = `
"import _examples_react_src_pages_about_tsx from \\"/examples/react/src/pages/about.tsx\\";
import _examples_react_src_pages_about_index_tsx from \\"/examples/react/src/pages/about/index.tsx\\";
import _examples_react_src_pages_components_tsx from \\"/examples/react/src/pages/components.tsx\\";
import _examples_react_src_pages_index_tsx from \\"/examples/react/src/pages/index.tsx\\";
import _examples_react_src_pages_$___all$_tsx from \\"/examples/react/src/pages/[...all].tsx\\";
import _examples_react_src_pages_blog_index_tsx from \\"/examples/react/src/pages/blog/index.tsx\\";
import _examples_react_src_pages_blog_$id$_tsx from \\"/examples/react/src/pages/blog/[id].tsx\\";
import _examples_react_src_pages_blog_today_index_tsx from \\"/examples/react/src/pages/blog/today/index.tsx\\";
import React from \\"react\\";

const routes = [{\\"path\\":\\"about\\",\\"element\\":React.createElement(_examples_react_src_pages_about_tsx),\\"children\\":[{\\"element\\":React.createElement(_examples_react_src_pages_about_index_tsx),\\"index\\":true}]},{\\"path\\":\\"components\\",\\"element\\":React.createElement(_examples_react_src_pages_components_tsx)},{\\"element\\":React.createElement(_examples_react_src_pages_index_tsx),\\"index\\":true},{\\"path\\":\\"*\\",\\"element\\":React.createElement(_examples_react_src_pages_$___all$_tsx)},{\\"path\\":\\"blog\\",\\"children\\":[{\\"element\\":React.createElement(_examples_react_src_pages_blog_index_tsx),\\"index\\":true},{\\"path\\":\\":id\\",\\"element\\":React.createElement(_examples_react_src_pages_blog_$id$_tsx)},{\\"path\\":\\"today\\",\\"children\\":[{\\"element\\":React.createElement(_examples_react_src_pages_blog_today_index_tsx),\\"index\\":true}]}]}];

export default routes;"
`;

exports[`Generate > React routes > routes 1`] = `
[
  {
    "children": [
      {
        "element": "/examples/react/src/pages/about/index.tsx",
        "index": true,
      },
    ],
    "element": "/examples/react/src/pages/about.tsx",
    "path": "about",
  },
  {
    "element": "/examples/react/src/pages/components.tsx",
    "path": "components",
  },
  {
    "element": "/examples/react/src/pages/index.tsx",
    "index": true,
  },
  {
    "element": "/examples/react/src/pages/[...all].tsx",
    "path": "*",
  },
  {
    "children": [
      {
        "element": "/examples/react/src/pages/blog/index.tsx",
        "index": true,
      },
      {
        "element": "/examples/react/src/pages/blog/[id].tsx",
        "path": ":id",
      },
      {
        "children": [
          {
            "element": "/examples/react/src/pages/blog/today/index.tsx",
            "index": true,
          },
        ],
        "path": "today",
      },
    ],
    "path": "blog",
  },
]
`;

exports[`Generate > Vue routes - async > client code 1`] = `
"import _examples_vue_src_pages_index_vue from \\"/examples/vue/src/pages/index.vue\\";

const routes = [{\\"path\\":\\"/about\\",\\"component\\":() => import('/examples/vue/src/pages/about.vue'),\\"children\\":[{\\"name\\":\\"about\\",\\"path\\":\\"\\",\\"component\\":() => import('/examples/vue/src/pages/about/index.vue'),\\"props\\":(route2)=>({query:route2.query.q})},{\\"path\\":\\":id\\",\\"component\\":() => import('/examples/vue/src/pages/about/[id].vue'),\\"children\\":[{\\"name\\":\\"about-id-more\\",\\"path\\":\\"more\\",\\"component\\":() => import('/examples/vue/src/pages/about/[id]/more.vue'),\\"props\\":true},{\\"name\\":\\"about-id-nested\\",\\"path\\":\\"nested\\",\\"component\\":() => import('/examples/vue/src/pages/about/[id]/nested.vue'),\\"props\\":true}],\\"props\\":true,\\"name\\":\\"about-user-id\\",\\"meta\\":{\\"requiresAuth\\":true}}],\\"props\\":true,\\"meta\\":{\\"lang\\":\\"yml\\"}},{\\"name\\":\\"components\\",\\"path\\":\\"/components\\",\\"component\\":() => import('/examples/vue/src/pages/components.vue'),\\"props\\":true,\\"meta\\":{\\"lang\\":\\"yaml\\"}},{\\"name\\":\\"homepage\\",\\"path\\":\\"/\\",\\"component\\":_examples_vue_src_pages_index_vue,\\"props\\":true,\\"meta\\":{\\"requiresAuth\\":false}},{\\"name\\":\\"all\\",\\"path\\":\\"/:all(.*)*\\",\\"component\\":() => import('/examples/vue/src/pages/[...all].vue'),\\"props\\":true},{\\"path\\":\\"/:sensor\\",\\"component\\":() => import('/examples/vue/src/pages/[sensor].vue'),\\"children\\":[{\\"name\\":\\"sensor-current\\",\\"path\\":\\"current\\",\\"component\\":() => import('/examples/vue/src/pages/[sensor]/current.vue'),\\"props\\":true}],\\"props\\":true},{\\"name\\":\\"blog\\",\\"path\\":\\"/blog\\",\\"component\\":() => import('/examples/vue/src/pages/blog/index.vue'),\\"props\\":true},{\\"name\\":\\"blog-id\\",\\"path\\":\\"/blog/:id\\",\\"component\\":() => import('/examples/vue/src/pages/blog/[id].vue'),\\"props\\":true,\\"meta\\":{\\"requiresAuth\\":false}},{\\"name\\":\\"blog-today\\",\\"path\\":\\"/blog/today\\",\\"component\\":() => import('/examples/vue/src/pages/blog/today/index.vue'),\\"props\\":true}];

export default routes;"
`;

exports[`Generate > Vue routes - async > routes 1`] = `
[
  {
    "children": [
      {
        "component": "/examples/vue/src/pages/about/index.vue",
        "customBlock": undefined,
        "name": "about",
        "path": "",
        "props": [Function],
      },
      {
        "children": [
          {
            "component": "/examples/vue/src/pages/about/[id]/more.vue",
            "customBlock": undefined,
            "name": "about-id-more",
            "path": "more",
            "props": true,
          },
          {
            "component": "/examples/vue/src/pages/about/[id]/nested.vue",
            "customBlock": undefined,
            "name": "about-id-nested",
            "path": "nested",
            "props": true,
          },
        ],
        "component": "/examples/vue/src/pages/about/[id].vue",
        "meta": {
          "requiresAuth": true,
        },
        "name": "about-user-id",
        "path": ":id",
        "props": true,
      },
    ],
    "component": "/examples/vue/src/pages/about.vue",
    "meta": {
      "lang": "yml",
    },
    "path": "/about",
    "props": true,
  },
  {
    "component": "/examples/vue/src/pages/components.vue",
    "meta": {
      "lang": "yaml",
    },
    "name": "components",
    "path": "/components",
    "props": true,
  },
  {
    "component": "/examples/vue/src/pages/index.vue",
    "meta": {
      "requiresAuth": false,
    },
    "name": "homepage",
    "path": "/",
    "props": true,
  },
  {
    "component": "/examples/vue/src/pages/[...all].vue",
    "customBlock": undefined,
    "name": "all",
    "path": "/:all(.*)*",
    "props": true,
  },
  {
    "children": [
      {
        "component": "/examples/vue/src/pages/[sensor]/current.vue",
        "customBlock": undefined,
        "name": "sensor-current",
        "path": "current",
        "props": true,
      },
    ],
    "component": "/examples/vue/src/pages/[sensor].vue",
    "customBlock": undefined,
    "path": "/:sensor",
    "props": true,
  },
  {
    "component": "/examples/vue/src/pages/blog/index.vue",
    "customBlock": undefined,
    "name": "blog",
    "path": "/blog",
    "props": true,
  },
  {
    "component": "/examples/vue/src/pages/blog/[id].vue",
    "meta": {
      "requiresAuth": false,
    },
    "name": "blog-id",
    "path": "/blog/:id",
    "props": true,
  },
  {
    "component": "/examples/vue/src/pages/blog/today/index.vue",
    "customBlock": undefined,
    "name": "blog-today",
    "path": "/blog/today",
    "props": true,
  },
]
`;

exports[`Generate > Vue routes - sync > client code 1`] = `
"import _examples_vue_src_pages_about_vue from \\"/examples/vue/src/pages/about.vue\\";
import _examples_vue_src_pages_about_index_vue from \\"/examples/vue/src/pages/about/index.vue\\";
import _examples_vue_src_pages_about_$id$_vue from \\"/examples/vue/src/pages/about/[id].vue\\";
import _examples_vue_src_pages_about_$id$_more_vue from \\"/examples/vue/src/pages/about/[id]/more.vue\\";
import _examples_vue_src_pages_about_$id$_nested_vue from \\"/examples/vue/src/pages/about/[id]/nested.vue\\";
import _examples_vue_src_pages_components_vue from \\"/examples/vue/src/pages/components.vue\\";
import _examples_vue_src_pages_index_vue from \\"/examples/vue/src/pages/index.vue\\";
import _examples_vue_src_pages_$___all$_vue from \\"/examples/vue/src/pages/[...all].vue\\";
import _examples_vue_src_pages_$sensor$_vue from \\"/examples/vue/src/pages/[sensor].vue\\";
import _examples_vue_src_pages_$sensor$_current_vue from \\"/examples/vue/src/pages/[sensor]/current.vue\\";
import _examples_vue_src_pages_blog_index_vue from \\"/examples/vue/src/pages/blog/index.vue\\";
import _examples_vue_src_pages_blog_$id$_vue from \\"/examples/vue/src/pages/blog/[id].vue\\";
import _examples_vue_src_pages_blog_today_index_vue from \\"/examples/vue/src/pages/blog/today/index.vue\\";

const routes = [{\\"path\\":\\"/about\\",\\"component\\":_examples_vue_src_pages_about_vue,\\"children\\":[{\\"name\\":\\"about\\",\\"path\\":\\"\\",\\"component\\":_examples_vue_src_pages_about_index_vue,\\"props\\":true},{\\"path\\":\\":id\\",\\"component\\":_examples_vue_src_pages_about_$id$_vue,\\"children\\":[{\\"name\\":\\"about-id-more\\",\\"path\\":\\"more\\",\\"component\\":_examples_vue_src_pages_about_$id$_more_vue,\\"props\\":true},{\\"name\\":\\"about-id-nested\\",\\"path\\":\\"nested\\",\\"component\\":_examples_vue_src_pages_about_$id$_nested_vue,\\"props\\":true}],\\"props\\":true,\\"name\\":\\"about-user-id\\",\\"meta\\":{\\"requiresAuth\\":true}}],\\"props\\":true,\\"meta\\":{\\"lang\\":\\"yml\\"}},{\\"name\\":\\"components\\",\\"path\\":\\"/components\\",\\"component\\":_examples_vue_src_pages_components_vue,\\"props\\":true,\\"meta\\":{\\"lang\\":\\"yaml\\"}},{\\"name\\":\\"homepage\\",\\"path\\":\\"/\\",\\"component\\":_examples_vue_src_pages_index_vue,\\"props\\":true,\\"meta\\":{\\"requiresAuth\\":false}},{\\"name\\":\\"all\\",\\"path\\":\\"/:all(.*)*\\",\\"component\\":_examples_vue_src_pages_$___all$_vue,\\"props\\":true},{\\"path\\":\\"/:sensor\\",\\"component\\":_examples_vue_src_pages_$sensor$_vue,\\"children\\":[{\\"name\\":\\"sensor-current\\",\\"path\\":\\"current\\",\\"component\\":_examples_vue_src_pages_$sensor$_current_vue,\\"props\\":true}],\\"props\\":true},{\\"name\\":\\"blog\\",\\"path\\":\\"/blog\\",\\"component\\":_examples_vue_src_pages_blog_index_vue,\\"props\\":true},{\\"name\\":\\"blog-id\\",\\"path\\":\\"/blog/:id\\",\\"component\\":_examples_vue_src_pages_blog_$id$_vue,\\"props\\":true,\\"meta\\":{\\"requiresAuth\\":false}},{\\"name\\":\\"blog-today\\",\\"path\\":\\"/blog/today\\",\\"component\\":_examples_vue_src_pages_blog_today_index_vue,\\"props\\":true}];

export default routes;"
`;

exports[`Generate > Vue routes - sync > routes 1`] = `
[
  {
    "children": [
      {
        "component": "/examples/vue/src/pages/about/index.vue",
        "customBlock": undefined,
        "name": "about",
        "path": "",
        "props": true,
      },
      {
        "children": [
          {
            "component": "/examples/vue/src/pages/about/[id]/more.vue",
            "customBlock": undefined,
            "name": "about-id-more",
            "path": "more",
            "props": true,
          },
          {
            "component": "/examples/vue/src/pages/about/[id]/nested.vue",
            "customBlock": undefined,
            "name": "about-id-nested",
            "path": "nested",
            "props": true,
          },
        ],
        "component": "/examples/vue/src/pages/about/[id].vue",
        "meta": {
          "requiresAuth": true,
        },
        "name": "about-user-id",
        "path": ":id",
        "props": true,
      },
    ],
    "component": "/examples/vue/src/pages/about.vue",
    "meta": {
      "lang": "yml",
    },
    "path": "/about",
    "props": true,
  },
  {
    "component": "/examples/vue/src/pages/components.vue",
    "meta": {
      "lang": "yaml",
    },
    "name": "components",
    "path": "/components",
    "props": true,
  },
  {
    "component": "/examples/vue/src/pages/index.vue",
    "meta": {
      "requiresAuth": false,
    },
    "name": "homepage",
    "path": "/",
    "props": true,
  },
  {
    "component": "/examples/vue/src/pages/[...all].vue",
    "customBlock": undefined,
    "name": "all",
    "path": "/:all(.*)*",
    "props": true,
  },
  {
    "children": [
      {
        "component": "/examples/vue/src/pages/[sensor]/current.vue",
        "customBlock": undefined,
        "name": "sensor-current",
        "path": "current",
        "props": true,
      },
    ],
    "component": "/examples/vue/src/pages/[sensor].vue",
    "customBlock": undefined,
    "path": "/:sensor",
    "props": true,
  },
  {
    "component": "/examples/vue/src/pages/blog/index.vue",
    "customBlock": undefined,
    "name": "blog",
    "path": "/blog",
    "props": true,
  },
  {
    "component": "/examples/vue/src/pages/blog/[id].vue",
    "meta": {
      "requiresAuth": false,
    },
    "name": "blog-id",
    "path": "/blog/:id",
    "props": true,
  },
  {
    "component": "/examples/vue/src/pages/blog/today/index.vue",
    "customBlock": undefined,
    "name": "blog-today",
    "path": "/blog/today",
    "props": true,
  },
]
`;
